pandas/core/reshape/merge.py

```diff
@@ -2411,18 +2411,30 @@ def _factorize_keys(
                 .combine_chunks()
                 .dictionary_encode()
             )
-            length = len(dc.dictionary)
 
             llab, rlab, count = (
-                pc.fill_null(dc.indices[slice(len_lk)], length)
+                pc.fill_null(dc.indices[slice(len_lk)], -1)
                 .to_numpy()
                 .astype(np.intp, copy=False),
-                pc.fill_null(dc.indices[slice(len_lk, None)], length)
+                pc.fill_null(dc.indices[slice(len_lk, None)], -1)
                 .to_numpy()
                 .astype(np.intp, copy=False),
                 len(dc.dictionary),
             )
+
+            if sort:
```

Comment:
This might be able to be done without converting to numpy

```
if sort:
    sorted_idx = pc.array_sort_indices(dc.indices, order=..., null_placement = ...)
    sorted_lab = pc.take(dc.indices, sorted_idx)
    llab, rlab = sorted_lab[slice(len_lk)], sorted_lab[slice(len_lk, None)]
```

Reply:
This would only work if the codes are already ordered correctly, which is not guaranteed